### DISCUSSION TITLE
Type-safe (via code generation) models and DAO helpers

### DISCUSSION THREAD

**Discussion by preslavrachev** [pocketbase/pocketbase#2697] [Polls]
I love using Pocketbase as a tiny Go framework for prototyping my side projects. There are only a few things that I think would make the experience even smoother for me, and one of those is working with the DAO. It is pretty solid as it is, but some of the interactions can get pretty repetitive, especially, when you start having to make calls across tables:

- fetch the column and handle possible errors
- fetch certain record(s) from the column and handle possible records
- select some data from each record, crossing your fingers that you selected the right types.
- repeat for all other tables

I noticed that Pocketbase is doing quite a stretch generating Go code for each database migration. This got me thinking, if PB can introspect what is in each table column, can't it also generate type-safe (well, as safe as you can go with a DAO) models and their respective methods? Thus, for example, getting some records from the `Orders` collection would be a single call returning an `Order` slice and error (or we can pass a pointer to a slice, I'm perfectly fine with it).

My alternative at the moment is using [ent](https://entgo.io/) as my sort of DAO that then goes in my `app` instance, alongside a reference to the original PB `app`. It works well, but having something like that as a first-class PB citizen would be fantastic.

**Comment by ganigeorgiev** [comment]
While it is possible to automate the generation of [Custom models](https://pocketbase.io/docs/custom-models/) as a prebuild step, it will be very difficult (_or at least I can't think of an easy way_) to make it work efficiently with the [`Record` hooks](https://pocketbase.io/docs/event-hooks/#record-crud-api-hooks) because at the moment they operate only with `models.Record`.

This is definitely something that can be improved but I don't think it can be implemented easily without breaking too much things.

One possible solution I guess could be to replace it with an interface and maybe embed the `models.Record` as part of the custom model struct but still it'll require somewhere to have a mapping with factories `collection -> customModelFactory()` that we'll use when retrieving data or triggering a hook event with that record.

This needs careful consideration and may require a lot of internal changes, and at least for the near future I'm not planning working on it as there are other more major [blocker tasks](https://github.com/orgs/pocketbase/projects/2/views/1) that needs to be resolved first (_not to mention that I'm already ~3months behind the initial schedule for [Presentator v3](https://github.com/presentator/presentator/issues/183)_).

**Reply by ganigeorgiev** [reply]
On second thought, there is maybe a way to implement it without breaking the existing functionality and even making it "optional/pluggable".
We can preserve the current `model.Record` and Record hooks (we need them also in case a collection is created/modified during runtime or the prebuilt executable is used), and instead create dynamically in some custom package, structs for each collection model that will operate on `*models.Record` with named getters and setters, eg.:

  ```go
  // GENERATED CODE - DO NOT MODIFY BY HAND
  //
  // (maybe also generate automatically a godoc...)
  package codegen

  type Article struct {
      record *models.Record
  }

  func (a *Article) Name() string {
      return a.record.GetString("name")
  }

  func (a *Article) SetName(name string) {
      return a.record.Set("name", name)
  }

  // ...
  ```

  To "resolve" the hooks issue, we can also autogenerate named hooks (eg. `OnArticleBeforeUpdateRequest()`) that will operate with their own `Article` based event in similar manner to the above.
  Those named hooks could be automatically triggered by the `OnRecord*` hooks like:

  ```go
  app.OnRecordBeforeUpdateRequest().PreAdd(func(e *core.RecordUpdateEvent) {
      return codegen.OnArticleBeforeUpdateRequest().Trigger(toArticleEvent(e))
  })
  
  // this will be triggered by OnRecordBeforeUpdateRequest
  codegen.OnArticleBeforeUpdateRequest().Add(func(e *codege.ArticleEvent) {
    e.Article.SetName("something...")
  })
  // etc.
  ```

It'll require some work and tests but I think it is doable. I've added it to the [roadmap](https://github.com/orgs/pocketbase/projects/2/views/1?pane=issue&itemId=30580834) but still for now remains a low priority and will consider it probably somewhere after the logs generalization.

**Reply by preslavrachev** [reply]
Thanks a lot for taking the time to even consider this feature! I perfectly understand your struggles with keeping deadlines and releases under control. That's why I opened this as a discussion and not as a direct feature request. I am aware that it's likely a big change, but I'm willing to bet that Pocketbase has the potential of becoming the missing Rails or Django for Go. With type-safe models (and some sort of built-in task scheduler), it's more than half-way there - if you want it to be, of course.

Anyways, thanks for the proposed solution. i will also spend some time thinking about how this could look like in practice. Pluggable sounds ideal to me. Take your time and good luck with the other issues.

**Reply by alexisvisco** [reply]
I have implemented a POC following what @ganigeorgiev is saying at this URL : https://github.com/alexisvisco/pocketpase-gen

(note the pocketpase typo 🤦, fortunately I will be on vacation at the end of the week )

For now it only generate getter and setter with some light documentation. 

Please add issues for things you want :)

**Comment by s00500** [comment]
Hey All,

UH I would love this, just thinking of how many bugs I already built cause of the strings in the record interface...
(I am really working on a side project with minimal time and then things like this are more likely to happen without type checks...)

I am currently giving the generator a shot, looks like exactly what I was searching for =D
I thing adding the name of the collection as a constant could also be cool still... might make a PR...
Also better ways to expand records... maybe also some shortcuts to record id or such (might be overkill)

Since the last progress on this is like 1 year old, any new thoughts on this in general ?

**Reply by ganigeorgiev** [reply]
A record "projection/proxy" interface was already introduced in v0.23.0-rc (it is in the `develop` branch). Go generate scripts or custom console command will be available after the v0.23.0 release based on that idea.

There is no public documentation for it yet, but I'll copy below part of the text I'm preparing for the new site:

----


The available `core.Record` and its helpers are usually the recommended way to interact with your data, but in case you want a typed access to your record fields you can create a helper struct that embeds `core.BaseRecordProxy` _(which implements the `core.RecordProxy` interface)_ and define your collection fields as getters and setters.

By implementing the `core.RecordProxy` interface you can use your custom struct as part of a
`RecordQuery` result like a regular record model. In addition, every DB change through the proxy
struct will trigger the corresponding record validations and hooks. This ensures that other parts of your app,
including 3rd party plugins, that doesn't know or use your custom struct will still work as expected.

Below is a sample `Article` record proxy implementation:

```go
// article.go
package main

import (
    "github.com/pocketbase/pocketbase/core"
    "github.com/pocketbase/pocketbase/tools/types"
)

// ensures that the Article struct satisfy the core.RecordProxy interface
var _ core.RecordProxy = (*Article)(nil)

type Article struct {
    core.BaseRecordProxy
}

func (a *Article) Title() string {
    return a.GetString("title")
}

func (a *Article) SetTitle(title string) {
    a.Set("title", title)
}

func (a *Article) Slug() string {
    return a.GetString("slug")
}

func (a *Article) SetSlug(slug string) {
    a.Set("slug", slug)
}

func (a *Article) Created() types.DateTime {
    return a.GetDateTime("created")
}

func (a *Article) Updated() types.DateTime {
    return a.GetDateTime("updated")
}
```

Accessing and modifying the proxy records is the same as for the regular records. Continuing with the
above `Article` example:

```go
func FindArticleBySlug(app core.App, slug string) (*Article, error) {
    article := &Article{}

    err := app.RecordQuery("articles").
        AndWhere(dbx.NewExp("LOWER(slug)={:slug}", dbx.Params{
            "slug": strings.ToLower(slug), // case insensitive match
        })).
        Limit(1).
        One(article)

    if err != nil {
        return nil, err
    }

    return article, nil
}

...

article, err := FindArticleBySlug(app, "example")
if err != nil {
    return err
}

// change the title
article.SetTitle("Lorem ipsum...")

// persist the change while also triggerring the original record validations and hooks
err = app.Save(article)
if err != nil {
    return err
}
```

If you have an existing `*core.Record` value you can also load it into your proxy using the `SetProxyRecord` method:

```go
// fetch regular record
record, err := app.FindRecordById("articles", "RECORD_ID")
if err != nil {
    return err
}

// load into proxy
article := &Article{}
article.SetProxyRecord(record)
```

**Reply by s00500** [reply]
Awesome 😍, just upgraded to 0.23-rc8 to give this a shot

Also super fast reply!

Noticed quite a few (very easily fixable) breaking changes, is there a place already where these are collected ? or is that for later ? :-)

**Reply by ganigeorgiev** [reply]
Note that v0.23 introduces a lot of changes and requires manual migration steps.

You can find some details about the changes in the `develop` [CHANGELOG.md](https://github.com/pocketbase/pocketbase/blob/develop/CHANGELOG.md) (starting from the bottom).

A minimal migration Go APIs guide is available in https://pocketbase.io/v023upgrade/go/.

For the JS and Dart SDKs, you'll have to refer to the specific repos CHANGELOGs:
- [JS SDK v0.22.0-rcX](https://github.com/pocketbase/js-sdk/blob/develop/CHANGELOG.md)
- [Dart SDK v0.19.0-rcX](https://github.com/pocketbase/dart-sdk/blob/develop/CHANGELOG.md)

A showcase of all new features will be available with the new documentation (_there are no ETAs_).

**Reply by s00500** [reply]
Awesome! thanks

**Reply by AbyssWaIker** [reply]
Is there a recommended way to handling json fields (for example, if I want to hide "collectionName" and "collectionId" from my custom models. Or change some other fields) or is it a low priority after 0.23?)

(I can manually override PublicExport and pass model like re.JSON(200, model.PublicExport())), but I would be very happy to discover that I'm missing something)

Thank you, in advance, for your response

**Reply by ganigeorgiev** [reply]
@AbyssWaIker I don't see how your question is related to the above discussion. For hiding fields based on some condition you can combine the `app.OnRecordEnrich` hook and the `record.Hide(fieldName)` method (there is no public documentation for it yet, but you can find an example in https://github.com/pocketbase/pocketbase/discussions/5369#discussioncomment-10524248).

If you have other questions, please consider opening a new Q&A discussion elaborating more on your use case what you are trying to do.

**Reply by AbyssWaIker** [reply]
Long story short - I'm defining custom route, where I accept request and after custom validation logic, save new record to database and return the record back as a response.

The relevant part of my hiccup is that before the rework custom model like the one from your example could be defined like this
```
 type Article struct {
    models.BaseModel
    Title string `db:"title" json:"title"`
    Slug string `db:"slug" json:"slug"`
    Created types.DateTime `db:"created" json:"created"`
    Updated types.DateTime `db:"updated" json:"updated"`
}
```
(With only the relevant fields present and json tag giving control over the name of fields). 
core.BaseRecordProxy does not give such fine grained control + shows unneded technical fields about the collection.


And OnRecordEnrich does not trigger when returning model like this
```
func MyRoute(re *core.RequestEvent) error {
///Custom logic
///...
	return re.JSON(http.StatusOK, newlyCreatedProxyRecord)
}
```

I was asking about that, but I understand why it may be a low priority (post 0.23) thing.

Or maybe extra fields with technical details about collection are shown because I'm returning it wrong. Or creating not optimally. Example how I create models 


```
func CreateMyCustomRecord(
	app core.App,
	//Params
	//...
) (*MyCustomRecord, error)
         var newRecord MyCustomRecord
        //I'm updating cache on every request, which may be a bad idea
	collection, err := (*app).FindCachedCollectionByNameOrId(TableName)
	if err != nil {
		return nil, err
	}
	newRecord.SetProxyRecord(core.NewRecord(collection))

	//Setters
	//...
	
	err = app.Save(newRecord)
	if err != nil {
		return nil, err
	}
	return &newRecord, nil
}
```
 
P.S. (And less relevant part is that I'm using using PreValidate/PostValidate/OnRecordEnrich/OnModelCreate/OnModelUpdate hooks to store values of Select fields in database not as strings, but as integers. And use them in go code as "enums" (uint8 type with overloaded MarshalJSON/UnmarshalJSON). Earlier json marshaling/unmarshaling would automatically convert  everything back and forth. Now I was hoping that I missed something like OnRecordEnrich for ProxyRecord to hijack. I understand, that I'm using it completely not intended way so I do not expect any support on this part)

P.P.SThank you again, for such quick answer

P.P.P.S I will be using `Record.Hide(fieldName)`

**Comment by ganigeorgiev** [comment]
I'm locking this discussion as it started to become too messy. If anyone has further questions please open a new Q&A discussion.
